### PR TITLE
chapel 1.25.1

### PR DESCRIPTION
**Pull request by Maxrimus**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
A note on the patches: We've included a number of end of file patch from an [upstream change](https://github.com/chapel-lang/chapel/pull/18820) and a testing branch to allow for the proper specification of our LLVM location and to help us get the M1 changes working. This patch should be removed next release as it is already in our upstream.

### CODE REVIEW

**Comment by Maxrimus** [comment]
Can anyone on the homebrew team tell us why most of the CI Tests were cancelled?

**Comment by bradcray** [comment]
Hi @carlocab —

I'm wondering whether you could give us some guidance or insight here since you've been so helpful in the past.  Specifically, you may recall that in October, we were wondering what it would take to build an ARM bottle for Chapel, and you ran a test for us, [which identified three issues](https://github.com/chapel-lang/chapel/issues/17910#issuecomment-950246280).

Since then, we've updated our formula to make use of homebrew GMP to address the third, and easiest issue (while also completing [the plan we created in response to your March review](https://github.com/chapel-lang/chapel/issues/17428) to use external packages when practical).  We have not worked on either of the other two M1-oriented failures, in part due to other competing priorities, but mostly due to not having an M1 Mac on which to test.  We're in the process of procuring one and hope to have it early in the new year.

It seems as though nothing has gotten worse w.r.t. M1 in our code base since 1.25.0, yet our 1.25.0 formula's PR didn't fail the M1 CI tests in this way.  Have the CI requirements changed since then, or did something in our formula (or PR tags?) change since then that would alter the way we are tested?

Bottom line:  Is there a way for this formula to proceed in spite of the M1 failures?

Thanks very much for any insight you can provide here,
-Brad

**Comment by Bo98** [comment]
In terms of the specific failures here, it's things that probably didn't have automatic detection before but do now. In short: it's things that probably shouldn't have been approved before but went unnoticed.

Both seem fixable however.

The qthreads one seems like an oversight upstream unless I'm missing something, given everything else in the repo is source-only but pre-built binaries are shipped for qthread. I don't see any reference anywhere that they would even be used. Can't the folder `third-party/qthread/qthread-src/installed` just be deleted?

chpl-venv looks fixable with `CHPL_PIP_INSTALL_PARAMS=--no-binary :all:`

**Comment by bradcray** [comment]
@Bo98 : Thanks very much for the information here, and the quick response, both of which are very helpful.  A few questions follow:

> didn't have automatic detection before but do now.

That's reassuring, and what I was hoping.  Is there a good way for semi-annual homebrew contributors like us to track changes to the CI pipeline?  We often find our homebrew release lags other releases, often due to our fault, but sometimes due to CI changes that we didn't anticipate and which catch us off-guard on release day.  I'm curious how we could head off these issues as we go.  For example, is there a way to engage the homebrew CI process on a regular basis or at a specific moment short of opening up a PR?  (or is it acceptable to open a PR a day or two before a release simply to see whether it's passing the CI?)

 > seems like an oversight upstream

Oh wow, I think you're right here and that none of us looked closely enough to recognize this.  I'm embarrassed.  I think we all assumed that the directory had been created by the build (which is what we'd normally expect) and hadn't noticed that it was pre-existing.  This also addresses the "I thought we'd stubbed this build out on ARM" question we were asking ourselves yesterday.  Thanks for pointing it out, and I think you're right.  Tagging @ronawho who added this and would know best.

> chpl-venv looks fixable with CHPL_PIP_INSTALL_PARAMS=--no-binary :all:

Thanks for that.  I'm not much of a Python guy and hadn't called in our Python experts yet because the first error seemed bigger and was stymying me.  We'll give it a shot.

Thanks again!

**Comment by ronawho** [comment]
>> The qthreads one seems like an oversight upstream unless I'm missing something, given everything else in the repo is source-only but prebuilt binaries are shipped for qthread. I don't see any reference anywhere that they would even be used. Can't the folder third-party/qthread/qthread-src/installed just be deleted?

> Oh wow, I think you're right here and that none of us looked closely enough to recognize this. I'm embarrassed. I think we all assumed that the directory had been created by the build (which is what we'd normally expect) and hadn't noticed that it was pre-existing. This also addresses the "I thought we'd stubbed this build out on ARM" question we were asking ourselves yesterday. Thanks for pointing it out, and I think you're right. Tagging @ronawho who added this and would know best.

Ah, yeah good catch. This is actually something I noticed while preparing to upgrade to the next qthreads version since the latest version doesn't have the files and when looking at the diff I didn't understand why they were there before. Sorry, I should have jumped to that being the issue earlier.

These files are certainly not needed. For now, I think we should just delete the the directory in our homebrew formulae and I'll make sure we upgrade qthreads versions before our next release. (The qthreads release process has been spotty in the past with extra generated files being added so they have an action items to automate their release creation instead of manually creating the tarball, so hopefully this won't be an issue going forward.)

**Comment by bradcray** [comment]
> I think we should just delete the the directory in our homebrew formulae

Thanks for verifying.  That's the approach Max is taking.

**Comment by carlocab** [comment]
Hi @bradcray -- I'm on holiday at the moment, so I haven't had the chance to look at this carefully (or read the other messages in this thread). I'll be back next week and will have a look then, but I think Bo will have sorted everything important here already.

**Comment by bradcray** [comment]
@carlocab — Thanks for replying even while on holiday and sorry for all the @-tags this week as we've struggled with this.  I think Bo's put us on the right path so hopefully you will return to nothing more to comment on here.  Thanks again for checking in.

File: Formula/chapel.rb [hunk 1]
```diff
@@ -60,9 +69,193 @@ def install
...
+      end
+      with_env(CHPL_LLVM: "none") do
+        system "util/test/checkChplInstall"
+      end
     end
     system bin/"chpl", "--print-passes", "--print-commands", libexec/"examples/hello.chpl"
   end
 end
+
+__END__
```
[the remark this answers is not in the document]
Comment (mppf) [reply Formula/chapel.rb:86]: Yes, they are in the work-in-progress PR https://github.com/chapel-lang/chapel/pull/18880 which I am expecting to complete soon along with some other changes that will improve portability

**Comment by mppf** [comment]
This PR Is ready to merge so please let us know if there's something we need to address. Thanks.

File: Formula/chapel.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (bradcray) [reply Formula/chapel.rb:86]: @SMillerDev: For a bit more context on the patches: We were caught by surprise by the CI checks in some cases with this PR because we don't have an M1 Mac in-house to do testing on (though we're working on addressing that) and didn't have problems with the M1 CI checks in our recent 1.25.0 release.  So until we opened up the PR and started seeing the failures due to the stricter M1 checks, we were unaware of the issues.  I think one of the others may have been due to the change in this formula to start using homebrew GMP rather than our bundled version, previously requested by @carlocab, which revealed some portability challenges due to different paths on the non-vanilla Mac builds.  As Michael says, we're working on incorporating changes into our `main` branch to avoid the need for these in future releases, but are hoping to proceed with this PR in the meantime to catch homebrew up with our other 1.25.1 releases.

[the remark this answers is not in the document]
Comment (SMillerDev) [reply Formula/chapel.rb:86]: That's okay, just wanted to ensure this wasn't meant to hang around in Homebrew forever.

File: Formula/chapel.rb [hunk 2]
```diff
@@ -14,36 +13,46 @@ class Chapel < Formula
     sha256 x86_64_linux: "7cf01213c80d41ed1b20451fd9f11b4b651a1b9a603729826a383e73475918b3"
   end
 
+  depends_on "gmp"
   depends_on "llvm@11"
   depends_on "python@3.9"
 
+  patch :DATA
```
Comment (SMillerDev) [inline Formula/chapel.rb:20]: Just one more thing:
```suggestion
  # can be removed after https://github.com/chapel-lang/chapel/pull/18880 merged
  patch :DATA
```

**Comment by mppf** [comment]
@SMillerDev - we've made the requested change, please let us know if there's anything else we need to do here, thanks.

**Review by SMillerDev — APPROVED** [review]
Thanks @Maxrimus and the rest of the team ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Comment by bradcray** [comment]
Thanks for helping us to wrestle this in, @Bo98, @SMillerDev, and @carlocab!

Given the trouble CI testing gave us this time around, we're curious:  Is there a homebrew-approved way to run a PR past homebrew's CI process without any intention of merging it?  For example, say that a week or 10 days before our next release, we wanted to make sure there weren't issues like this lurking in the homebrew CI using a rough draft of our release code.  Is there a way we could open a PR to preview the CI process before the day of the release without having it be accidentally merged?  I looked around a bit in the documentation for things like this, but didn't see any guidance one way or the other.

Thanks again.

**Comment by SMillerDev** [comment]
Yeah, just mark it as a prerelease upstream and ask a maintainer for a `do not merge` label
